src/runtime/virtcontainers/clh.go

```diff
@@ -1261,37 +1406,71 @@ func (clh *cloudHypervisor) bootVM(ctx context.Context) error {
 		}
 		clh.Logger().WithField("body", string(bodyBuf)).Debug("VM config")
 	}
-	_, err := cl.CreateVM(ctx, clh.vmconfig)
-	if err != nil {
-		return openAPIClientError(err)
-	}
 
-	info, err := clh.vmInfo()
-	if err != nil {
-		return err
-	}
+	if !clh.config.BootFromTemplate {
+		_, err := cl.CreateVM(ctx, clh.vmconfig)
+		if err != nil {
+			return openAPIClientError(err)
+		}
 
-	clh.Logger().Debugf("VM state after create: %#v", info)
+		info, err := clh.vmInfo()
+		if err != nil {
+			return err
+		}
 
-	if info.State != clhStateCreated {
-		return fmt.Errorf("VM state is not 'Created' after 'CreateVM'")
-	}
+		clh.Logger().Debugf("VM state after create: %#v", info)
 
-	clh.Logger().Debug("Booting VM")
-	_, err = cl.BootVM(ctx)
-	if err != nil {
-		return openAPIClientError(err)
-	}
+		if info.State != clhStateCreated {
+			return fmt.Errorf("VM state is not 'Created' after 'CreateVM'")
+		}
 
-	info, err = clh.vmInfo()
-	if err != nil {
-		return err
-	}
+		clh.Logger().Debug("Booting VM")
+		_, err = cl.BootVM(ctx)
+		if err != nil {
+			return openAPIClientError(err)
+		}
+
+		info, err = clh.vmInfo()
+		if err != nil {
+			return err
+		}
 
-	clh.Logger().Debugf("VM state after boot: %#v", info)
+		clh.Logger().Debugf("VM state after boot: %#v", info)
+
+		if info.State != clhStateRunning {
+			return fmt.Errorf("VM state is not 'Running' after 'BootVM'")
+		}
+	} else {
+		clh.Logger().Info("Restoring VM from template")
 
-	if info.State != clhStateRunning {
-		return fmt.Errorf("VM state is not 'Running' after 'BootVM'")
+		vmStoreDir := filepath.Join(clh.config.VMStorePath, clh.id)
+		vmStatePath := filepath.Join(vmStoreDir, stateJson)
+		err := os.Symlink(clh.config.DevicesStatePath, vmStatePath)
+		if err != nil {
+			return err
+		}
+
+		data, errJSON := json.MarshalIndent(clh.vmconfig, "", "\t")
+		if errJSON != nil {
+			return errJSON
+		}
+		vmConfigPath := filepath.Join(vmStoreDir, configJson)
+		err = os.WriteFile(vmConfigPath, data, 0600)
```

Comment:
Thanks for changing this. It might be worth making this a top-level `const` for extra clarity.